Tutorials physics simulation terminology (#120)
ignitionrobotics/ign-physics: tutorials/06-physics-simulation-concepts.md

\page physicsconcepts Ignition Physics simulation concepts

This tutorial introduces simulation concepts that are used in Ignition Physics.
As the SDF model is used to describe simulation scenarios in Ignition, the
Ignition Physics simulation concepts are described closely related to the SDF model.

## Physics simulation versus animation

Dynamics simulators for robotics is more challenging than the ones used for
animating virtual characters due to these aspects:
- The forces, torques, frictions, etc. are not crucial in animation since the
law of physics can be violated.
- Real-time requirements and physical reality can be less constraining for
purely visual display.
- Several challenges for simulating complex physical structures, such as
humanoid robot:
  - Numerical instability in computing inverse kinematics, fluid flows, etc.
    prevents correct physics simulation.
  - Contact forces between bodies are hard to model due to their discontinuity
    characteristics, especially in the soft body case, may result in unrealistic
    contacts or physically unfeasible contact forces.

> Ignition adopts SDF structure to describe not only the visual element but also

Reviewer: I recommend `SDFormat` instead of `SDF`, since there are some other acronyms that use `SDF`.

      ...
    </physics>

    <scene>
      ...
    </scene>

    <model name="box">
      ...
    </model>

    <model name="sphere">
      ...
    </model>

    <light name="spotlight">
      ...
    </light>
  </world>
</sdf>
```

The next level of the root element `sdf` specifying the SDF format version
is the `world` element, which encapsulates entire simulation scenarios
description including these elements:
- `physics`: specifies the type of physics engines for dynamic simulation (in
  our case they are `TPE` or `dartsim`).
- `scene`: specifies the view of the environment.
- `model`: defines a complex robot structure or an object. Note there can be
  multiple `model` inside `world` tag, which enables multi-body simulation.
- `light`: describes light source in simulation scenarios.

### Physics simulation concepts corresponding to the SDF model

Looking closer to the `model` tag, an example structure defining a simple car is
described below with comments:

```xml
<model name="car">
    <pose>0 0 0.5 0 0 0</pose> <!--specifies where the car pose in world frame-->
    <static>false</static> <!--specifies whether the care is static (unmovable) or not-->

>     <link name="chassic"> <!--specifies a component of a car with physics properties-->

Reviewer: `chassic` -> `chassis`

humanoid robot:
  - Numerical instability in computing inverse kinematics, fluid flows, etc.
    prevents correct physics simulation.
  - Contact forces between bodies are hard to model due to their discontinuity
    characteristics, especially in the soft body case, may result in unrealistic
    contacts or physically unfeasible contact forces.

Ignition adopts SDFormat structure to describe not only the visual element but
also the dynamic physics aspects. To get more in-depth of what you can define
the environment in the SDFormat file, please refer to this
[SDFormat specification](http://sdformat.org/spec?ver=1.7&elem=sdf).
For a comprehensive tutorial for constructing your robot model as SDFormat,
please refer to [this tutorial](https://ignitionrobotics.org/docs/dome/building_robot).

## Physics concepts in Ignition Gazebo simulation

First, please see [Understand the GUI tutorial](https://ignitionrobotics.org/docs/dome/gui)
for an overview of Ignition Gazebo GUI. In this tutorial, we will show how to
manipulate and visualize some physics aspects using interesting
models powered by Ignition Physics on Ignition Gazebo.

> ### Differential Car

Reviewer 2: ```suggestion
### Differential Drive
```

> This model demonstrates how we can control simulated differential cars depending

Reviewer 2: ```suggestion
This model demonstrates how we can control simulated differential drive on cars depending
```

on physics engines and visualize the collision concept. Please download the

> Differential Car model to your home folder by:

Reviewer 2: ```suggestion
Differential drive simulation world demo to your home folder by:
```

links' poses relative to their parent link by selecting the
corresponding link on the model tree:

<img src="https://user-images.githubusercontent.com/18066876/97189395-d45ef380-17a4-11eb-9c29-cf33b5b184c4.gif"/>

Note that using the model tree as shown in the above gif, we can view the
parameters and properties such as `visual` or `collision` of each link or joint
in the model. For more detail about the kinematics of the model tree, see this
[tutorial](http://sdformat.org/tutorials?tut=spec_model_kinematics&cat=specification&).

#### Visualizing collision

One of the most wonderful features of physics simulation is the capability to
simulate collision. We can do a fun experiment like this: while the `vehicle_blue`
car is moving in a circle, we move the `vehicle_green` to be on `vehicle_blue`'s
upcoming path. We will see the blue car will push the green car!

<img src="https://user-images.githubusercontent.com/18066876/97190565-28b6a300-17a6-11eb-9ed5-0d6d50c15ad4.gif"/>

To see where these collision parameters are set in SDFormat and how it works,
please see this [tutorial](http://sdformat.org/tutorials?tut=spec_shapes&cat=specification&).

### Lift Drag

> This model shows how joint force, torque, and pressure are supported in Ignition Physics.

Reviewer 2: ```suggestion
This demo world shows how joint force, torque, and pressure are supported in Ignition Physics.
```

> Please download the Lift Drag model to your home folder by:

Reviewer 2: ```suggestion
Please download the Lift Drag demo to your home folder by:
```

```bash
wget https://raw.githubusercontent.com/ignitionrobotics/ign-gazebo/main/examples/worlds/lift_drag.sdf -P ~
```

> Like above, please start the Lift Drag model on Ignition Gazebo by:

Reviewer 2: Nit: a model refers to a single car or robot in a world, in this case we are using a entire simulation world with light, grid, plane and all those prebuilt into the sdf file. 
```suggestion
Like above, please start the Lift Drag demo world on Ignition Gazebo by:
```

```

Then please press Play button to start the simulation. We stops exerting torque by:

```bash
ign topic -t "/model/lift_drag_demo_model/joint/rod_1_joint/cmd_force" -m ignition.msgs.Double  -p "data: 0.0"
```

You will see the cube drops due to no lift force from support torque on the rod,
and the blades will stop after some time due to friction.

<img src="https://user-images.githubusercontent.com/18066876/99107715-8d675f80-25e6-11eb-91f3-90f83bbc93e9.gif"/>

The command applies a constant torque to the rotor rod, together with
the mechanism to compute the upward/downward lift and drag force due to the
wind pressure simulation supported by Ignition Physics, the cube is lifted.
For more detail, please see the [LiftDrag.cc](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/systems/lift_drag/LiftDrag.cc)
code.

### Buoyancy

This model shows how buoyancy is supported in Ignition Physics. This world
contains the following three models:

  1. submarine: A simple submarine model that floats in place.
  2. submarine_sinking: A simple submarine model that is not buoyant and sinks.
  3. submarine_buoyant: A simple submarine model that is buoyant and floats.

> Please download the Buoyancy model to your home folder by:

Reviewer 2: ```suggestion
Please download the Buoyancy demo to your home folder by:
```

```bash
wget https://raw.githubusercontent.com/ignitionrobotics/ign-gazebo/ign-gazebo4/examples/worlds/buoyancy.sdf -P ~
```

> Like above, please start the Buoyancy model on Ignition Gazebo by:

Reviewer 2: same as ^

links' poses relative to their parent link by selecting the
corresponding link on the model tree:

<img src="https://user-images.githubusercontent.com/18066876/97189395-d45ef380-17a4-11eb-9c29-cf33b5b184c4.gif"/>

Note that using the model tree as shown in the above gif, we can view the
parameters and properties such as `visual` or `collision` of each link or joint
in the model. For more detail about the kinematics of the model tree, see this [tutorial](http://sdformat.org/tutorials?tut=spec_model_kinematics&cat=specification&).

### Visualizing collision

One of the most wonderful features of physics simulation is the capability to
simulate collision. We can do a fun experiment like this: while the `vehicle_blue`
car is moving in a circle, we move the `vehicle_green` to be on `vehicle_blue`'s
upcoming path. We will see the blue car will push the green car!

<img src="https://user-images.githubusercontent.com/18066876/97190565-28b6a300-17a6-11eb-9ed5-0d6d50c15ad4.gif"/>

To see where these collision parameters are set and how it works, please see
this [tutorial](http://sdformat.org/tutorials?tut=spec_shapes&cat=specification&).

> ## Physics concepts in SDFormat

Reviewer 2: I think the next few sections about SDFormat isn't closely related to ign-physics and most of them can be referred back to sdformat documentation. I think it'd be nicer if we can focus more on the existing demo worlds.

Some example worlds that might be worth demonstrating here:
1. [lift_drag](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/worlds/lift_drag.sdf) shows how joint force, torque and pressure is supported in ign-physics, source code see https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/systems/lift_drag/LiftDrag.cc
2. [video_record_dbl_pendulum](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/worlds/video_record_dbl_pendulum.sdf) shows how inertia and gravity affect the object movement by showing free swing of the pendulum 
3. [multicopter_velocity_control](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/worlds/multicopter_velocity_control.sdf) shows how ign-physics supports gravity, actuators and inertia to control object velocity, source code https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/systems/multicopter_control/MulticopterVelocityControl.cc
4. [trisphere_cycle_wheel_slip](https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/examples/worlds/trisphere_cycle_wheel_slip.sdf.) This is a fun one. This one shows the slip compliance and collision supported by ign-physics with source code https://github.com/ignitionrobotics/ign-gazebo/blob/ign-gazebo4/src/systems/wheel_slip/WheelSlip.cc

Author: Sorry I messed up a bit :), I exclude the Wheel Slip for now because it reports some errors. I also include Bouyancy because it seems important.
